src/bindings/swig/python/plugins/dns/src/main/dns_plugin.py

```diff
@@ -0,0 +1,86 @@
+import kdb
+import socket
+
+
+"""
+Usage:
+> sudo kdb mount file.ini /python python script=/path/to/filter_script.py
+> sudo kdb meta-set user:/python/my_hostname check/dns ''
+> sudo kdb set user:/python/my_hostname www.libelektra.org
+"""
+
+META_DNS_NAME = "check/dns"
+
+
+def get_ipv4_by_hostname(hostname) -> bool:
+	return bool([
+		i[4][0]  # address
+		for i in socket.getaddrinfo(hostname, 0)
+		if i[0] is socket.AddressFamily.AF_INET and i[1] is socket.SocketKind.SOCK_RAW
+	])
+
+
+def check_key(key: kdb.Key):
+	print(key, key.value)
+	# we only check if Meta META_DNS_NAME is set
+	if m := key.getMeta(META_DNS_NAME):
+		print(m.value)
+		if key.value != '':
+			try:
+				return get_ipv4_by_hostname(key.value)
+			except Exception as e:
+				return False
+
+	return True
+
+
+class ElektraPlugin(object):
+	def __init__(self):
+		pass
+
+	def open(self, config: kdb.KeySet, errorKey):
+		"""
+		returns:
+		#  - nil or 0: no error
+		#  - -1      : error during initialization
+		"""
+		return 0
+
+	def get(self, returned: kdb.KeySet, parentKey: kdb.Key):
+		"""
+		#  - nil or 1 : on success
+		#  -        0 : OK but nothing was to do
+		#  -       -1 : failure
+		"""
+		mod = "system:/elektra/modules/python"
+		if parentKey.name == mod:
+			returned.append(kdb.Key(mod, kdb.KEY_VALUE, "contract below"))
+			returned.append(kdb.Key(mod+"/infos", kdb.KEY_VALUE, "contract below"))
+			returned.append(kdb.Key(mod+"/infos/placements", kdb.KEY_VALUE, "postgetstorage presetstorage"))
+			return 1
+		return 1
+
+	def set(self, returned: kdb.KeySet, parentKey: kdb.Key):
+		"""
+		#  - nil or 1 : on success
+		#           0 : on success with no changed keys in database
+		#          -1 : failure
+		"""
+
+		for k in returned:
+			if not check_key(k):
+				print(f"couldn't validate key {k}")
```

Comment:
How exactly should we do that? Is there documentation that explains that? I found [this](https://doc.libelektra.org/api/latest/html/group__plugin.html#ga3da3bdb0f41710adda9eee3d7adac9ff) but it doesn't seem to apply here.
My proto-solution is  something like: `parentKey.setMeta("warning", f"Couldn't resolve domain name for key(s): {warning_list}")` in get() and error in set() respectively.

Reply:
AFAIK there sadly is no python library function for adding errors. There was some recent work for C++, but it is not yet accessible via Python.

The C macros eventually call into the functions in [`errors.c`](https://github.com/ElektraInitiative/libelektra/blob/master/src/libs/elektra/errors.c). You'll have to replicate the parts the behaviour appropriate for your case (i.e. set all the same meta keys to the correct values). The Java binding also has [`Key.setError` and `Key.setWarning`](https://github.com/ElektraInitiative/libelektra/blob/master/src/bindings/jna/libelektra/src/main/java/org/libelektra/Key.java), which serve the same purpose as the C macros, but may be easier to understand.

Reply:
PS. It would be awesome, if you created Python versions of Java's `Key.setError` and `Key.setWarning`. I'm no expert with SWIG, but I think you have to add the Python functions into this section of `kdb.i`:

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/bindings/swig/python/kdb.i#L151-L157

Then the functions should be accessible on the `Key` class like the `setMeta` that is defined there.

Reply:
> How exactly should we do that?

Errors/warnings are basically meta-data added to the parentKey/errorKey.

> Is there documentation that explains that?

doc/dev/error-handling.md (especially section "Metadata")

ideally you build something like described in

https://www.libelektra.org/tutorials/language-bindings

It could be within your plugin for today (so that we can merge it after the deadline) and probably later you could add it to kdb.i like suggested by @kodebach.

Reply:
I added the error handling in accordance with doc/dev/error-handling.md in [814325e](https://github.com/ElektraInitiative/libelektra/pull/4344/commits/814325e8b25051ac1d2ed48fac702f2126765653) within the plugin.

One last thing: I wasn't able to find information about the error number apart from the fact that it should be unique. Is there a list with all error code so that I can choose one that is not already used elsewhere?
Or does it just have to be unique on module-level? Right now I defined the number 101 for our case.

Reply:
The error codes are defined here

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/libs/elektra/errors.c#L17-L34

(Java has them defined [here](https://github.com/ElektraInitiative/libelektra/blob/master/src/bindings/jna/libelektra/src/main/java/org/libelektra/ErrorCode.java))

As you can see, each error number has a fixed description. Please make sure, you use this description. Only the `reason` metakey is meant to describe the individual error.

Also please set all the metakeys `number`, `description`, `module`, `file`, `line`, `mountpoint`, `configfile` and `reason`. The `mountpoint` and `configfile` keys can simply be set to the `keyName` and `keyString` of the `parentKey`. For `file` and `line` you can simply set `"unknown"` and `0` for now.

PS. I think in your case `VALIDATION_SEMANTIC` should be the appropriate error type.

Reply:
Okay thanks,
With `keyString` do you mean `parentKey.value` and with `keyName`  `parentKey.name`?

Reply:
Explanation to the error codes is in doc/dev/error-categorization.md